monitoring/monitoring.go

```diff
@@ -0,0 +1,61 @@
+package monitoring
+import (
+	"google.golang.org/grpc/codes"
+)
+
+
+type RpcType string
+
+const (
+	Unary RpcType = "unary"
+	Streaming RpcType = "streaming"
+)
+
+
+// RpcMonitor is a per-RPC datastructure
+type RpcMonitor interface {
+	// ReceivedMessage is called on every stream message received by the monitor.
+	ReceivedMessage()
+
+	// SentMessage is called on every stream message sent by the monitor.
+	SentMessage()
+
+	// Handled is called whenever the RPC handling completes (with OK or AppError).
+	Handled(code codes.Code)
+
+	// Errored is called whenever the RPC failed due to RPC-layer errors.
+	Erred(err error)
+}
+
+
+// ServerMonitor allocates new per-RPC monitors on the server side.
```

Comment:
I didn't want to bring confusion as it seems the prevelant opinion in the bug thread is that Tracing data structures should be different from monitoring ones: https://github.com/grpc/grpc-go/issues/240#issuecomment-120210769